File: frontend/src/services/orderDetailsService.ts
import axios, { AxiosError } from "axios";

/**
 * DTOs for Order Details Modernized Form
 */

// State persistence DTO
export interface OrderDetailStatePersistenceDto {
  itemsListSelectedKey?: string;
  observationsListSelectedKey?: string;
  blocksListSelectedKey?: string;
  invoicesListSelectedKey?: string;
  autoLoadItemsChecked?: boolean;
  autoLoadObservationsChecked?: boolean;
  autoLoadBlocksChecked?: boolean;
  autoLoadInvoicesChecked?: boolean;
}

// Order Item DTO
export interface OrderItemDto {
  orderNumber: string;
  productCode: string;
  sequentialId: number;
  orderedQuantity?: number;
  invoicedQuantity?: number;
  destinedQuantity?: number;
  committedQuantity?: number;
  status?: string;
  priceValue?: number;
  balanceValue?: number;
  unitValue?: number;
  discount?: number;
  paymentCondition?: string;
  priceTable?: string;
  fiscalGroupPricing?: string;
  fiscalOperationPricing?: string;
  fiscalGroupDelivery?: string;
  fiscalOperationDelivery?: string;
  earlyDate?: string;
  lateDate?: string;
  baseDate?: string;
  productShortDescription?: string;
  productLongDescription?: string;
}

// Order Observation DTO
export interface OrderObservationDto {
  orderNumber: string;
  sequentialId: number;
  operationTypeDescription?: string;
  stateRegistration?: string;
  name?: string;
  address?: string;
  mrh?: string;
  city?: string;
  state?: string;
  municipality?: string;
  invoiceText?: string;
  freeText?: string;
}

// Order Block DTO
export interface OrderBlockDto {
  orderNumber: string;
  lineNumber?: string;
  sequentialId: number;
  blockDescription?: string;
  status?: string;
  statusDate?: string;
  message?: string;
  blockTypeId?: string;
}

// Order Invoice DTO
export interface OrderInvoiceDto {
  invoiceCode: string;
  series: string;
  orderNumber?: string;
  clientCode?: string;
  establishment?: string;
  factoryCode?: string;
  statusCode?: string;
  invoiceType?: string;
  emissionDate?: string;
  merchandiseExitDate?: string;
  icmsBaseValue?: number;
  icmsValue?: number;
  ipiValue?: number;
  icmsAliquotValue?: number;
  netWeight?: number;
  grossWeight?: number;
  discountValue?: number;
  totalValue?: number;
  totalInvoicedUnits?: number;
  volumeQuantity?: number;
  transportWay?: string;
  transportDescription?: string;
  additionalDiscountValue?: number;
  qualityDescription?: string;
  currencyCode?: string;
  factoryDescription?: string;
  clientCnpj?: string;
  clientRazaoSocial?: string;
  clientTradeName?: string;
  transportCode?: string;
}

// Main Order Details DTO
export interface OrderDetailsDto {
  orderNumber: string;
  clientCnpj: string;
  clientRazaoSocial: string;
  inclusionDate?: string;
  establishment?: string;
  clientCode?: string;
  transportDescription?: string;
  lastUpdateDate?: string;
  clientOrderNumber?: string;
  post?: string;
  totalOrderValue?: number;
  commercialConditionDescription?: string;
  billingStatusDescription?: string;
  orderStatusCode?: string;
  orderStatusDescription?: string;
  requiredDate?: string;
  items: OrderItemDto[];
  observations: OrderObservationDto[];
  blocks: OrderBlockDto[];
  invoices: OrderInvoiceDto[];
  autoLoadItems?: boolean;
  autoLoadObservations?: boolean;
  autoLoadBlocks?: boolean;
  autoLoadInvoices?: boolean;
  statePersistence?: OrderDetailStatePersistenceDto;
}

/**
 * Error response from API
 */
export interface ApiErrorResponse {
  message: string;
  code?: string;
  details?: any;
  timestamp?: string;
  traceId?: string;
}

/**
 * API base URL
 */
const API_BASE_URL = "/api/OrderDetails";

/**
 * Helper to handle API errors and throw user-friendly messages.
 */
function handleApiError(error: any): never {
  if (axios.isAxiosError(error)) {
    const err = error as AxiosError<ApiErrorResponse>;
    if (err.response && err.response.data && err.response.data.message) {
      throw new Error(err.response.data.message);
    }
    if (err.message) {
      throw new Error(err.message);
    }
  }
  throw new Error("Ocorreu um erro inesperado ao processar sua solicitação.");
}

/**
 * Get all order details, including client info, items, observations, blocks, invoices, and persisted state.
 */
export async function getOrderDetails(orderNumber: string): Promise<OrderDetailsDto> {
  try {
    const response = await axios.get<OrderDetailsDto>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Update and get the latest list of items for a given order.
 */
export async function updateOrderItems(orderNumber: string): Promise<OrderItemDto[]> {
  try {
    const response = await axios.get<OrderItemDto[]>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/items`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Update and get the latest list of observations for a given order.
 */
export async function updateOrderObservations(orderNumber: string): Promise<OrderObservationDto[]> {
  try {
    const response = await axios.get<OrderObservationDto[]>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/observations`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Update and get the latest list of blocks for a given order.
 */
export async function updateOrderBlocks(orderNumber: string): Promise<OrderBlockDto[]> {
  try {
    const response = await axios.get<OrderBlockDto[]>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/blocks`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Update and get the latest list of invoices for a given order.
 */
export async function updateOrderInvoices(orderNumber: string): Promise<OrderInvoiceDto[]> {
  try {
    const response = await axios.get<OrderInvoiceDto[]>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/invoices`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Export the specified list type (items, observations, blocks, invoices) for a given order to an Excel file.
 * Returns a Blob for download.
 */
export async function exportListToExcel(orderNumber: string, listType: "items" | "observations" | "blocks" | "invoices"): Promise<Blob> {
  try {
    const response = await axios.get(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/export/${listType}`, {
      responseType: "blob",
    });
    return response.data as Blob;
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Persist the state of list view positions and checkbox values for a given order.
 */
export async function persistState(orderNumber: string, state: OrderDetailStatePersistenceDto): Promise<void> {
  try {
    await axios.post(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/state`, state);
  } catch (error) {
    handleApiError(error);
  }
}

/**
 * Retrieve the persisted state of list view positions and checkbox values for a given order.
 */
export async function getPersistedState(orderNumber: string): Promise<OrderDetailStatePersistenceDto> {
  try {
    const response = await axios.get<OrderDetailStatePersistenceDto>(`${API_BASE_URL}/${encodeURIComponent(orderNumber)}/state`);
    return response.data;
  } catch (error) {
    handleApiError(error);
  }
}

export default {
  getOrderDetails,
  updateOrderItems,
  updateOrderObservations,
  updateOrderBlocks,
  updateOrderInvoices,
  exportListToExcel,
  persistState,
  getPersistedState,
};

File: OrderDetailsService/OrderDetailsService.Application/Services/OrderDetailsService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using OrderDetailsService.Application.DTOs;
using OrderDetailsService.Application.Interfaces;
using OrderDetailsService.Domain.Entities;
using OrderDetailsService.Infrastructure.Repositories;

namespace OrderDetailsService.Application.Services
{
    /// <summary>
    /// Service responsible for all business logic related to the modernized Order Details form,
    /// including retrieval, update, export, and state persistence for order details, items, observations, blocks, and invoices.
    /// </summary>
    public class OrderDetailsService : IOrderDetailsService
    {
        private readonly IOrderDetailsRepository _orderDetailsRepository;
        private readonly IExcelExportService _excelExportService;
        private readonly IStatePersistenceService _statePersistenceService;

        public OrderDetailsService(
            IOrderDetailsRepository orderDetailsRepository,
            IExcelExportService excelExportService,
            IStatePersistenceService statePersistenceService)
        {
            _orderDetailsRepository = orderDetailsRepository;
            _excelExportService = excelExportService;
            _statePersistenceService = statePersistenceService;
        }

        /// <inheritdoc />
        public async Task<OrderDetailsDto> GetOrderDetailsAsync(string orderNumber)
        {
            if (string.IsNullOrWhiteSpace(orderNumber))
                throw new ArgumentException("Order number must be provided.", nameof(orderNumber));

            // Retrieve order and all related data
            var order = await _orderDetailsRepository.GetOrderWithDetailsAsync(orderNumber);
            if (order == null)
                throw new KeyNotFoundException($"Order '{orderNumber}' not found.");

            // Retrieve persisted state (positions, checkboxes)
            var state = await _statePersistenceService.GetPersistedStateAsync(orderNumber);

            // Map to DTO
            var dto = MapOrderToDto(order, state);

            return dto;
        }

        /// <inheritdoc />
        public async Task<OrderItemDto[]> UpdateOrderItemsAsync(string orderNumber)
        {
            var items = await _orderDetailsRepository.GetOrderItemsAsync(orderNumber);
            return items.Select(MapOrderItemToDto).ToArray();
        }

        /// <inheritdoc />
        public async Task<OrderObservationDto[]> UpdateOrderObservationsAsync(string orderNumber)
        {
            var observations = await _orderDetailsRepository.GetOrderObservationsAsync(orderNumber);
            return observations.Select(MapOrderObservationToDto).ToArray();
        }

        /// <inheritdoc />
        public async Task<OrderBlockDto[]> UpdateOrderBlocksAsync(string orderNumber)
        {
            var blocks = await _orderDetailsRepository.GetOrderBlocksAsync(orderNumber);
            return blocks.Select(MapOrderBlockToDto).ToArray();
        }

        /// <inheritdoc />
        public async Task<OrderInvoiceDto[]> UpdateOrderInvoicesAsync(string orderNumber)
        {
            var invoices = await _orderDetailsRepository.GetOrderInvoicesAsync(orderNumber);
            return invoices.Select(MapOrderInvoiceToDto).ToArray();
        }

        /// <inheritdoc />
        public async Task<byte[]> ExportListToExcelAsync(string orderNumber, string listType)
        {
            return await _excelExportService.ExportListToExcelAsync(orderNumber, listType);
        }

        /// <inheritdoc />
        public async Task PersistStateAsync(string orderNumber, OrderDetailStatePersistenceDto state)
        {
            await _statePersistenceService.PersistStateAsync(orderNumber, state);
        }

        /// <inheritdoc />
        public async Task<OrderDetailStatePersistenceDto> GetPersistedStateAsync(string orderNumber)
        {
            return await _statePersistenceService.GetPersistedStateAsync(orderNumber);
        }

        /// <inheritdoc />
        public async Task<ExportToSapResponse> ExportToSapAsync(string orderNumber)
        {
            // Simulação de exportação para SAP.
            // Em um cenário real, aqui seria feita a integração com o SAP (ex: via RFC, API REST, etc).
            // Para este exemplo, apenas retorna sucesso e um número fictício de documento SAP.

            if (string.IsNullOrWhiteSpace(orderNumber))
                throw new ArgumentException("Order number must be provided.", nameof(orderNumber));

            // Aqui você pode adicionar lógica real de integração SAP.
            // Exemplo: var sapResult = await _sapIntegrationService.ExportOrderAsync(orderNumber);

            // Simulação de sucesso:
            var response = new ExportToSapResponse
            {
                Success = true,
                Message = $"Pedido {orderNumber} exportado com sucesso para o SAP.",
                SapDocumentNumber = $"SAP{orderNumber.PadLeft(10, '0')}"
            };

            return await Task.FromResult(response);
        }

        #region Mapping Methods

        private static OrderDetailsDto MapOrderToDto(Order order, OrderDetailStatePersistenceDto state)
        {
            return new OrderDetailsDto
            {
                OrderNumber = order.OrderNumber,
                ClientCnpj = order.ClientCnpj,
                ClientRazaoSocial = order.ClientRazaoSocial,
                InclusionDate = order.InclusionDate,
                Establishment = order.Establishment,
                ClientCode = order.ClientCode,
                TransportDescription = order.TransportDescription,
                LastUpdateDate = order.LastUpdateDate,
                ClientOrderNumber = order.ClientOrderNumber,
                Post = order.Post,
                TotalOrderValue = order.TotalOrderValue,
                CommercialConditionDescription = order.CommercialConditionDescription,
                BillingStatusDescription = order.BillingStatusDescription,
                OrderStatusCode = order.OrderStatusCode,
                OrderStatusDescription = order.OrderStatusDescription,
                RequiredDate = order.RequiredDate,
                Items = order.Items?.Select(MapOrderItemToDto).ToList() ?? new List<OrderItemDto>(),
                Observations = order.Observations?.Select(MapOrderObservationToDto).ToList() ?? new List<OrderObservationDto>(),
                Blocks = order.Blocks?.Select(MapOrderBlockToDto).ToList() ?? new List<OrderBlockDto>(),
                Invoices = order.Invoices?.Select(MapOrderInvoiceToDto).ToList() ?? new List<OrderInvoiceDto>(),
                AutoLoadItems = order.AutoLoadItems,
                AutoLoadObservations = order.AutoLoadObservations,
                AutoLoadBlocks = order.AutoLoadBlocks,
                AutoLoadInvoices = order.AutoLoadInvoices,
                StatePersistence = state ?? new OrderDetailStatePersistenceDto()
            };
        }

        private static OrderItemDto MapOrderItemToDto(OrderItem item)
        {
            return new OrderItemDto
            {
                OrderNumber = item.OrderNumber,
                ProductCode = item.ProductCode,
                SequentialId = item.SequentialId,
                OrderedQuantity = item.OrderedQuantity,
                InvoicedQuantity = item.InvoicedQuantity,
                DestinedQuantity = item.DestinedQuantity,
                CommittedQuantity = item.CommittedQuantity,
                Status = item.Status,
                PriceValue = item.PriceValue,
                BalanceValue = item.BalanceValue,
                UnitValue = item.UnitValue,
                Discount = item.Discount,
                PaymentCondition = item.PaymentCondition,
                PriceTable = item.PriceTable,
                FiscalGroupPricing = item.FiscalGroupPricing,
                FiscalOperationPricing = item.FiscalOperationPricing,
                FiscalGroupDelivery = item.FiscalGroupDelivery,
                FiscalOperationDelivery = item.FiscalOperationDelivery,
                EarlyDate = item.EarlyDate,
                LateDate = item.LateDate,
                BaseDate = item.BaseDate,
                ProductShortDescription = item.ProductShortDescription,
                ProductLongDescription = item.ProductLongDescription
            };
        }

        private static OrderObservationDto MapOrderObservationToDto(OrderObservation obs)
        {
            return new OrderObservationDto
            {
                OrderNumber = obs.OrderNumber,
                SequentialId = obs.SequentialId,
                OperationTypeDescription = obs.OperationTypeDescription,
                StateRegistration = obs.StateRegistration,
                Name = obs.Name,
                Address = obs.Address,
                Mrh = obs.Mrh,
                City = obs.City,
                State = obs.State,
                Municipality = obs.Municipality,
                InvoiceText = obs.InvoiceText,
                FreeText = obs.FreeText
            };
        }

        private static OrderBlockDto MapOrderBlockToDto(OrderBlock block)
        {
            return new OrderBlockDto
            {
                OrderNumber = block.OrderNumber,
                LineNumber = block.LineNumber,
                SequentialId = block.SequentialId,
                BlockDescription = block.BlockDescription,
                Status = block.Status,
                StatusDate = block.StatusDate,
                Message = block.Message,
                BlockTypeId = block.BlockTypeId
            };
        }

        private static OrderInvoiceDto MapOrderInvoiceToDto(OrderInvoice invoice)
        {
            return new OrderInvoiceDto
            {
                InvoiceCode = invoice.InvoiceCode,
                Series = invoice.Series,
                OrderNumber = invoice.OrderNumber,
                ClientCode = invoice.ClientCode,
                Establishment = invoice.Establishment,
                FactoryCode = invoice.FactoryCode,
                StatusCode = invoice.StatusCode,
                InvoiceType = invoice.InvoiceType,
                EmissionDate = invoice.EmissionDate,
                MerchandiseExitDate = invoice.MerchandiseExitDate,
                IcmsBaseValue = invoice.IcmsBaseValue,
                IcmsValue = invoice.IcmsValue,
                IpiValue = invoice.IpiValue,
                IcmsAliquotValue = invoice.IcmsAliquotValue,
                NetWeight = invoice.NetWeight,
                GrossWeight = invoice.GrossWeight,
                DiscountValue = invoice.DiscountValue,
                TotalValue = invoice.TotalValue,
                TotalInvoicedUnits = invoice.TotalInvoicedUnits,
                VolumeQuantity = invoice.VolumeQuantity,
                TransportWay = invoice.TransportWay,
                TransportDescription = invoice.TransportDescription,
                AdditionalDiscountValue = invoice.AdditionalDiscountValue,
                QualityDescription = invoice.QualityDescription,
                CurrencyCode = invoice.CurrencyCode,
                FactoryDescription = invoice.FactoryDescription,
                ClientCnpj = invoice.ClientCnpj,
                ClientRazaoSocial = invoice.ClientRazaoSocial,
                ClientTradeName = invoice.ClientTradeName,
                TransportCode = invoice.TransportCode
            };
        }

        #endregion
    }
}